java.nio.file.NoSuchFileException

i put AMI code init in my spring boot application,but got error:

this is my code:

```
fun main(args: Array<String>) {
    runApplication<BPBXApplication>(*args)
    try {
        PBXFactory.init(AsteriskSettings())

        val asteriskPbx = PBXFactory.getActivePBX() as AsteriskPBX
        val agiEnabled = asteriskPbx.createAgiEntryPoint()
        if (!agiEnabled) {
            throw RuntimeException("AGI not Enabled !")
        }
        while (true) {
            try {
                Thread.sleep(60000)
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }
        }
        
        AMIUtil.checkConf();

    } catch (e: IOException) {
        e.printStackTrace()
    } catch (e: AuthenticationFailedException) {
        e.printStackTrace()
    } catch (e: TimeoutException) {
        e.printStackTrace()
    }
}

```



this is my log detail:



> 2024-01-24T11:03:55.502+08:00  INFO 3978653 --- [           main] com.byteee.bpbx.BPBXApplicationKt        : Started BPBXApplicationKt in 8.402 seconds (process running for 9.259)
> 2024-01-24T11:03:55.537+08:00  INFO 3978653 --- [           main] o.a.m.internal.ManagerConnectionImpl     : Connecting to 127.0.0.1:5038
> 2024-01-24T11:03:55.699+08:00  INFO 3978653 --- [           main] org.reflections.Reflections              : Reflections took 121 ms to scan 1 urls, producing 33 keys and 224 values
> 2024-01-24T11:03:55.831+08:00  INFO 3978653 --- [:AsyncEventPump] o.a.manager.internal.AsyncEventPump      : starting
> 2024-01-24T11:03:55.833+08:00  INFO 3978653 --- [:AsyncEventPump] o.a.m.internal.ManagerConnectionImpl     : Connected via Asterisk Call Manager/5.0.0
> 2024-01-24T11:03:55.843+08:00  INFO 3978653 --- [           main] o.a.m.internal.ManagerConnectionImpl     : Successfully logged in
> 2024-01-24T11:03:55.843+08:00  INFO 3978653 --- [           main] o.a.m.internal.ManagerConnectionImpl     : Got asterisk protocol identifier version Asterisk Call Manager/5.0.0
> 2024-01-24T11:03:55.845+08:00  INFO 3978653 --- [           main] o.a.m.internal.ManagerConnectionImpl     : Determined Asterisk version: Asterisk 16
> 2024-01-24T11:03:55.858+08:00  INFO 3978653 --- [           main] org.asteriskjava.util.ReflectionUtil     : >d:/www/bpbx-api/bpbx-api-0.0.1-SNAPSHOT.jar/
> 2024-01-24T11:03:55.858+08:00 ERROR 3978653 --- [           main] org.asteriskjava.util.ReflectionUtil     : java.nio.file.NoSuchFileException: d:/www/bpbx-api/bpbx-api-0.0.1-SNAPSHOT.jar
> 
> java.nio.file.NoSuchFileException: d:/www/bpbx-api/bpbx-api-0.0.1-SNAPSHOT.jar
>         at java.base/sun.nio.fs.UnixException.translateToIOException(UnixException.java:92) ~[na:na]
>         at java.base/sun.nio.fs.UnixFileAttributeViews$Basic.readAttributes(UnixFileAttributeViews.java:55) ~[na:na]
>         at java.base/sun.nio.fs.UnixFileSystemProvider.readAttributes(UnixFileSystemProvider.java:148) ~[na:na]
>         at java.base/sun.nio.fs.LinuxFileSystemProvider.readAttributes(LinuxFileSystemProvider.java:99) ~[na:na]
>         at java.base/java.util.zip.ZipFile$CleanableResource.<init>(ZipFile.java:709) ~[na:na]
>         at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:243) ~[na:na]
>         at java.base/java.util.zip.ZipFile.<init>(ZipFile.java:172) ~[na:na]
>         at java.base/java.util.jar.JarFile.<init>(JarFile.java:347) ~[na:na]
>         at java.base/java.util.jar.JarFile.<init>(JarFile.java:318) ~[na:na]
>         at java.base/java.util.jar.JarFile.<init>(JarFile.java:257) ~[na:na]
>         at org.asteriskjava.pbx.internal.core.CoherentManagerConnection.sendAction(CoherentManagerConnection.java:293) ~[asterisk-java-3.39.0.jar!/:na]
>         at org.asteriskjava.pbx.PBXFactory.getActivePBX(PBXFactory.java:10) ~[asterisk-java-3.39.0.jar!/:na]
>         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:na]
>         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77) ~[na:na]
>         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:na]
>         at java.base/java.lang.reflect.Method.invoke(Method.java:568) ~[na:na]
>         at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:91) ~[bpbx-api-0.0.1-SNAPSHOT.jar:0.0.1-SNAPSHOT]
>         at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:53) ~[bpbx-api-0.0.1-SNAPSHOT.jar:0.0.1-SNAPSHOT]
>         at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:58) ~[bpbx-api-0.0.1-SNAPSHOT.jar:0.0.1-SNAPSHOT]
> 
> Exception in thread "main" java.lang.reflect.InvocationTargetException
>         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
>         at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
>         at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
>         at java.base/java.lang.reflect.Method.invoke(Method.java:568)
>         at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:91)
>         at org.springframework.boot.loader.launch.Launcher.launch(Launcher.java:53)
>         at org.springframework.boot.loader.launch.JarLauncher.main(JarLauncher.java:58)
> Caused by: java.lang.ExceptionInInitializerError
>         at org.asteriskjava.pbx.PBXFactory.getActivePBX(PBXFactory.java:10)
>         at org.asteriskjava.pbx.PBXFactory.init(PBXFactory.java:18)
>         at com.byteee.bpbx.BPBXApplicationKt.main(BPBXApplication.kt:23)
>         ... 7 more
> Caused by: java.lang.RuntimeException: java.lang.RuntimeException: The requested event type of class org.asteriskjava.pbx.asterisk.wrap.events.NewChannelEvent+isn't known by class org.asteriskjava.pbx.internal.core.CoherentEventFactory
>         at org.asteriskjava.pbx.internal.core.AsteriskPBX.<init>(AsteriskPBX.java:60)
>         at org.asteriskjava.pbx.internal.core.AsteriskPBX.<clinit>(AsteriskPBX.java:37)
>         ... 10 more
> Caused by: java.lang.RuntimeException: The requested event type of class org.asteriskjava.pbx.asterisk.wrap.events.NewChannelEvent+isn't known by class org.asteriskjava.pbx.internal.core.CoherentEventFactory
>         at org.asteriskjava.pbx.internal.core.FilteredManagerListenerWrapper.<init>(FilteredManagerListenerWrapper.java:32)
>         at org.asteriskjava.pbx.internal.core.ListenerManager.addListener(ListenerManager.java:41)
>         at org.asteriskjava.pbx.internal.core.CoherentManagerEventQueue.addListener(CoherentManagerEventQueue.java:235)
>         at org.asteriskjava.pbx.internal.core.CoherentManagerConnection.addListener(CoherentManagerConnection.java:416)
>         at org.asteriskjava.pbx.internal.core.LiveChannelManager.<init>(LiveChannelManager.java:67)
>         at org.asteriskjava.pbx.internal.core.AsteriskPBX.<init>(AsteriskPBX.java:56)
>         ... 11 more

--- a/src/main/java/org/asteriskjava/pbx/internal/core/CoherentEventFactory.java
+++ b/src/main/java/org/asteriskjava/pbx/internal/core/CoherentEventFactory.java
@@ -35,7 +35,7 @@ public class CoherentEventFactory {
     // static initialiser
     static {
 
-        Set<Class<ManagerEvent>> knownClasses = ReflectionUtil.loadClasses("org.asteriskjava.pbx.asterisk.wrap.events",
+        Set<Class<? extends ManagerEvent>> knownClasses = ReflectionUtil.loadClasses("org.asteriskjava.pbx.asterisk.wrap.events",
                 ManagerEvent.class);
 
         ClassLoader classLoader = Thread.currentThread().getContextClassLoader();

--- a/src/main/java/org/asteriskjava/util/ReflectionUtil.java
+++ b/src/main/java/org/asteriskjava/util/ReflectionUtil.java
@@ -16,6 +16,8 @@
  */
 package org.asteriskjava.util;
 
+import org.reflections.Reflections;
+
 import java.io.File;
 import java.io.IOException;
 import java.lang.reflect.Constructor;
@@ -30,6 +32,7 @@ import java.util.*;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
+import java.util.stream.Collectors;
 
 /**
  * Utility class that provides helper methods for reflection that is used by the
@@ -257,35 +260,21 @@ public class ReflectionUtil {
     private static final Log logger = LogFactory.getLog(ReflectionUtil.class);
 
     /**
-     * find and all non abstract classes that implement/extend
-     * baseClassOrInterface in the package packageName
+     * Find all non-abstract classes in the given package that
+     * implement/extend the provided type.
      *
-     * @param packageName
-     * @param baseClassOrInterface
-     * @return
+     * @param packageName the package to search
+     * @param baseClassOrInterface the supertype or interface to filter by
+     * @return a Set of types that implement or extend the provided type
      */
-    @SuppressWarnings("unchecked")
-    public static <T> Set<Class<T>> loadClasses(String packageName, Class<T> baseClassOrInterface) {
-        Set<Class<T>> result = new HashSet<>();
+    public static <T> Set<Class<? extends T>> loadClasses(String packageName, Class<T> baseClassOrInterface) {
+        Set<Class<? extends T>> result = new Reflections(packageName)
+            .getSubTypesOf(baseClassOrInterface)
+            .stream()
+            .filter(c -> !Modifier.isAbstract(c.getModifiers()))
+            .collect(Collectors.toSet());
 
-        try {
-            Set<String> classNames = getClassNamesFromPackage(packageName);
-            ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
-            for (String className : classNames) {
-                try {
-                    Class<?> clazz = classLoader.loadClass(packageName + "." + className);
-                    if (!Modifier.isAbstract(clazz.getModifiers()) && baseClassOrInterface.isAssignableFrom(clazz)) {
-                        result.add((Class<T>) clazz);
-                    }
-                } catch (Throwable e) {
-                    logger.error(e, e);
-                }
-
-            }
-            logger.info("Loaded " + result.size());
-        } catch (Exception e) {
-            logger.error(e, e);
-        }
+        logger.info("Loaded " + result.size());
 
         return result;
     }

--- a/src/main/java/org/asteriskjava/util/ReflectionUtil.java
+++ b/src/main/java/org/asteriskjava/util/ReflectionUtil.java
@@ -33,6 +33,7 @@ import java.util.concurrent.ConcurrentHashMap;
 import java.util.jar.JarEntry;
 import java.util.jar.JarFile;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 /**
  * Utility class that provides helper methods for reflection that is used by the
@@ -268,13 +269,16 @@ public class ReflectionUtil {
      * @return a Set of types that implement or extend the provided type
      */
     public static <T> Set<Class<? extends T>> loadClasses(String packageName, Class<T> baseClassOrInterface) {
-        Set<Class<? extends T>> result = new Reflections(packageName)
-            .getSubTypesOf(baseClassOrInterface)
-            .stream()
-            .filter(c -> !Modifier.isAbstract(c.getModifiers()))
-            .collect(Collectors.toSet());
+        Set<Class<? extends T>> result =
+            Stream.concat(
+                Stream.of(baseClassOrInterface),
+                    new Reflections(packageName)
+                        .getSubTypesOf(baseClassOrInterface)
+                        .stream())
+                .filter(c -> !Modifier.isAbstract(c.getModifiers()))
+                .collect(Collectors.toSet());
 
-        logger.info("Loaded " + result.size());
+        logger.info("Loaded asd" + result.size());
 
         return result;
     }

